### PR TITLE
Twitter pics

### PR DESCRIPTION
**Pull request by chaseemory**
Lets Tweet!

### CODE REVIEW

File: cogs/Twitter.py [hunk 1]
```diff
@@ -0,0 +1,693 @@
...
+        servers_channels_raw = await db.get_twitter_sorting(
+            session, tag_list=tags, guild_list=servers
+        )
+        servers_dict = {server._guild: set() for server in servers_channels_raw}
+
+        # get channels into dict per server
+        for server in servers_channels_raw:
+            # just one of each, so we use a set since
+            # multiple hashtags can map to the same channel
+            servers_dict[server._guild].add(server._channel)
```
Comment (noahkw) [inline cogs/Twitter.py:589]: Directly construct a set of channels here (`.channel`)

File: db.py [hunk 2]
```diff
@@ -91,6 +95,84 @@ async def get_botw_nomination(session, guild_id, member_id):
     return result[0] if result else None
 
 
+async def get_twitter_settings(session, guild_id):
+    statement = select(TwtSetting).where(TwtSetting._guild == guild_id)
+    result = (await session.execute(statement)).all()
+
+    return [r for (r,) in result]
```
Comment (noahkw) [inline db.py:102]:
```suggestion
async def get_twitter_settings(session, guild_id):
    statement = select(TwtSetting).where(TwtSetting._guild == guild_id)
    result = (await session.execute(statement)).first()

    return result[0] if result else None
```

This also makes the multiple [0]'s above obsolete

File: db.py [hunk 3]
```diff
@@ -91,6 +95,84 @@ async def get_botw_nomination(session, guild_id, member_id):
...
+async def get_twitter_filters(session, _filter=None, guild_id=None):
+    if guild_id and _filter:
+        statement = select(TwtFilter).where(
+            (TwtFilter._filter == _filter) & (TwtFilter._guild == guild_id)
+        )
+    elif guild_id:
+        statement = select(TwtFilter).where(TwtFilter._guild == guild_id)
+    result = (await session.execute(statement)).all()
+
+    return [r for (r,) in result]
```
Comment (noahkw) [inline db.py:148]:
```suggestion
async def get_twitter_filters(session, filter_=None, guild_id=None):
    if guild_id and filter_:
        statement = select(TwtFilter).where(
            (TwtFilter._filter == filter_) & (TwtFilter._guild == guild_id)
        )
    elif guild_id:
        statement = select(TwtFilter).where(TwtFilter._guild == guild_id)
    result = (await session.execute(statement)).all()

    return [r for (r,) in result]
```

File: db.py [hunk 4]
```diff
@@ -91,6 +95,84 @@ async def get_botw_nomination(session, guild_id, member_id):
...
+        )
+    elif account_id:
+        statement = select(TwtAccount).where(TwtAccount.account_id == account_id)
+    elif _guild:
+        statement = select(TwtAccount).where(TwtAccount._guild == _guild)
+    else:
+        statement = select(TwtAccount)
+    result = (await session.execute(statement)).all()
+
+    return [r for (r,) in result]
```
Comment (noahkw) [inline db.py:118]:
```suggestion
async def get_twitter_accounts(session, account_id=None, guild_id=None):
    if guild_id and account_id:
        statement = select(TwtAccount).where(
            (TwtAccount.account_id == account_id) & (TwtAccount._guild == guild_id)
        )
    elif account_id:
        statement = select(TwtAccount).where(TwtAccount.account_id == account_id)
    elif guild_id:
        statement = select(TwtAccount).where(TwtAccount._guild == guild_id)
    else:
        statement = select(TwtAccount)
    result = (await session.execute(statement)).all()

    return [r for (r,) in result]
```

File: cogs/Twitter.py [hunk 5]
```diff
@@ -0,0 +1,693 @@
...
+            bitrate_and_urls = [
+                (variant["bitrate"], variant["url"])
+                for variant in variants
+                if "bitrate" in variant
+            ]
+            video_url = max(bitrate_and_urls, key=lambda x: x[0])[1]
+            video_filename = video_url.split("/")[-1].split("?")[0]
+
+            async with self.session.get(video_url) as response:
+                video_bytes = BytesIO(await response.read())
```
Comment (noahkw) [inline cogs/Twitter.py:632]: Pass session instead of creating one here

File: cogs/Twitter.py [hunk 6]
```diff
@@ -0,0 +1,693 @@
...
+    async def streaming(self):
+        self.feed = self.client.stream.statuses.filter.post(follow=self.gen_accounts)
+
+        async with self.feed as stream:
+            self.stream_thing = stream
+            async for tweet in self.stream_thing:
+                if on_tweet_with_media(tweet):
+                    logger.info(
+                        f"stream processing @{tweet.user.screen_name} - {tweet.id_str}"
+                    )
```
Comment (noahkw) [inline cogs/Twitter.py:95]:
```suggestion
                    logger.debug(
                        f"stream processing @{tweet.user.screen_name} - {tweet.id_str}"
                    )
```

File: cogs/Twitter.py [hunk 7]
```diff
@@ -0,0 +1,693 @@
...
+
+    @auto_help
+    @commands.group(
+        name="twitter",
+        aliases=["twt"],
+        brief="Automatically post twitter pics to specific channels based on tags and accounts",
+        invoke_without_command=True,
+    )
+    async def twitter(self, ctx):
+        await ctx.send_help(self.twitter)
```
Comment (noahkw) [inline cogs/Twitter.py:106]:
```suggestion
    )
    async def twitter(self, ctx):
        if not ctx.invoked_subcommand:
            await ctx.send_help(self.twitter)
```

Change all the other groups too.

File: cogs/Twitter.py [hunk 8]
```diff
@@ -0,0 +1,693 @@
...
+
+    @twitter.command(brief="Prints Twitter Stats")
+    @commands.is_owner()
+    async def stats(self, ctx):
+        accts_followed = len(self.gen_accounts)
+        time_start = time.time()
+        await self.client.api.statuses.show.get(
+            id=1230927030163628032, include_entities=False, trim_user=True
```
Comment (noahkw) [inline cogs/Twitter.py:114]: Magic number? Better as a constant. Though I am not sure what this does.

File: cogs/Twitter.py [hunk 9]
```diff
@@ -0,0 +1,693 @@
...
+                    server_settings[0].enabled = True
+            else:
+                server_settings = TwtSetting(
+                    _guild=ctx.guild.id,
+                    default_channel=0,
+                    use_group_channel=False,
+                    enabled=True,
+                )
+                session.add(server_settings)
+                logger.info(f"{ctx.guild.name} enabled twitter")
```
Comment (noahkw) [inline cogs/Twitter.py:155]:
```suggestion
                logger.info(f"{ctx.guild.name} ({ctx.guild.id}) enabled twitter")
```
Best to change other occurrences too. Just the name isn't very useful

File: cogs/Twitter.py [hunk 10]
```diff
@@ -0,0 +1,693 @@
...
+    @twitter_enabled()
+    async def post(self, ctx, tweet_url: commands.clean_content):
+        """
+        Will post individual tweet using sorting methods as implemented with hashtags by server
+
+        Example usage:
+        {prefix}twitter post https://twitter.com/Mias_Other_Half/status/1379979650605137928
+        """
+        tweet_urls = re.finditer(self.URL_REGEX, tweet_url)
+        tweet_ids = [_tweet_url.group(4) for _tweet_url in tweet_urls]
```
Comment (noahkw) [inline cogs/Twitter.py:187]:
```suggestion
        tweet_ids = [tweet_url_.group(4) for tweet_url_ in tweet_urls]
```

File: cogs/Twitter.py [hunk 11]
```diff
@@ -0,0 +1,693 @@
...
+
+    @twitter.group(
+        name="configure",
+        aliases=["config"],
+        brief="Set server specific details, default_channel and role",
+        invoke_without_command=True,
+    )
+    @twitter_enabled()
+    async def configure(self, ctx):
+        await ctx.send_help(self.set)
```
Comment (noahkw) [inline cogs/Twitter.py:207]: See the comment above

File: cogs/Twitter.py [hunk 12]
```diff
@@ -0,0 +1,693 @@
...
+    async def default_channel(self, ctx, channel: discord.TextChannel = None):
+        """
+        Set a default channel for when multiple channels match a tweet.
+        Such as #group or #OT5 or #<Group_Name>
+        Leave unset to have tweet posted in all matching channels
+        Leave blank to clear default channel
+        """
+        async with self.bot.Session() as session:
+            server_settings = await db.get_twitter_settings(session, ctx.guild.id)
+            server_settings[0].default_channel = channel.id if channel else 0
```
Comment (noahkw) [inline cogs/Twitter.py:225]:
```suggestion
            server_settings[0].default_channel = channel.id if channel else None
```

File: cogs/Twitter.py [hunk 13]
```diff
@@ -0,0 +1,693 @@
...
+        tag_conditioned = hashtag.split("#")[-1]
+        # check if hashtag already exists
+        async with self.bot.Session() as session:
+            twitter_sorting = await db.get_twitter_sorting(
+                session, hashtag=tag_conditioned, guild_id=ctx.guild.id
+            )
+            if twitter_sorting:
+                # twitter_sorting[0].channel = channel.id
+                await db.delete_twitter_sorting(session, tag_conditioned, ctx.guild.id)
+                logger.info(f"{ctx.guild.name} updated {tag_conditioned} to #{channel}")
```
Comment (noahkw) [inline cogs/Twitter.py:266]:
```suggestion
                logger.debug(f"{ctx.guild.name} updated {tag_conditioned} to #{channel}")
```
Most of these would be pretty spammy. Not sure whether I'd want to see them in my info logs

File: cogs/Twitter.py [hunk 14]
```diff
@@ -0,0 +1,693 @@
...
+            twitter_filter = await db.get_twitter_filters(
+                session, _filter=_filter, guild_id=ctx.guild.id
+            )
+            if twitter_filter:
+                await ctx.send(f'Filter "{_filter}" already exists')
+            else:
+                twitter_filter = TwtFilter(_guild=ctx.guild.id, _filter=_filter)
+                logger.info(f"{ctx.guild.name} added filter - {_filter}")
+                session.add(twitter_filter)
+                await session.commit()
```
Comment (noahkw) [inline cogs/Twitter.py:301]:
```suggestion
    async def add_filter(self, ctx, filter_: commands.clean_content):
        """
        Add a filter to your tweet stream, any tweets containing these filters will not be posted.

        Example usage:
        `{prefix}twitter add filter 프리뷰`
        """
        async with self.bot.Session() as session:
            # check if filter already exists
            twitter_filter = await db.get_twitter_filters(
                session, filter_=filter_, guild_id=ctx.guild.id
            )
            if twitter_filter:
                await ctx.send(f'Filter `{filter_}` already exists')
            else:
                twitter_filter = TwtFilter(_guild=ctx.guild.id, filter_=filter_)
                logger.info(f"{ctx.guild.name} ({ctx.guild.id}) added filter - {filter_}")
                session.add(twitter_filter)
                await session.commit()
```

File: cogs/Twitter.py [hunk 15]
```diff
@@ -0,0 +1,693 @@
...
+
+        Example usage:
+        `{prefix}twitter add account thinkB329`
+        """
+        # in case they enter it with the @ character
+        account_conditioned = account.split("@")[-1]
+        try:
+            account = await self.get_account(ctx, account=account_conditioned)
+        except TwitterAccountError:
+            return None
```
Comment (noahkw) [inline cogs/Twitter.py:319]: Better to show an error to the user.

File: cogs/Twitter.py [hunk 16]
```diff
@@ -0,0 +1,693 @@
...
+
+        Example usage:
+        `{prefix}twitter remove account thinkB329`
+        """
+        # in case they enter it with the @ character
+        account_conditioned = account.split("@")[-1]
+        try:
+            account = await self.get_account(ctx, account=account_conditioned)
+        except TwitterAccountError:
+            return None
```
Comment (noahkw) [inline cogs/Twitter.py:385]: Send an error msg here too

File: cogs/Twitter.py [hunk 17]
```diff
@@ -0,0 +1,693 @@
...
+
+        Example usage:
+        `{prefix}twitter remove filter 프리뷰`
+        """
+        async with self.bot.Session() as session:
+            await db.delete_twitter_filters(
+                session, _filter=_filter, guild_id=ctx.guild.id
+            )
+            logger.info(f"{ctx.guild.name} removed filter - {_filter}")
+            await session.commit()
```
Comment (noahkw) [inline cogs/Twitter.py:416]:
```suggestion
    async def remove_filter(self, ctx, filter_: commands.clean_content):
        """
        Remove a filter to your tweet stream, any tweets containing these filters will not be posted.

        Example usage:
        `{prefix}twitter remove filter 프리뷰`
        """
        async with self.bot.Session() as session:
            await db.delete_twitter_filters(
                session, filter_=filter_, guild_id=ctx.guild.id
            )
            logger.info(f"{ctx.guild.name} removed filter - {filter_}")
            await session.commit()
```

File: db.py (hunk 4 above)
Comment (noahkw) [inline db.py:118]:
```suggestion
async def get_twitter_accounts(session, account_id=None, guild_id=None):
    if guild_id and account_id:
        statement = select(TwtAccount).where(
            (TwtAccount.account_id == account_id) & (TwtAccount._guild == guild_id)
        )
    elif account_id:
        statement = select(TwtAccount).where(TwtAccount.account_id == account_id)
    elif guild_id:
        statement = select(TwtAccount).where(TwtAccount._guild == guild_id)
    else:
        statement = select(TwtAccount)
    result = (await session.execute(statement)).all()

    return [r for (r,) in result]
```

File: cogs/Twitter.py [hunk 18]
```diff
@@ -0,0 +1,693 @@
...
+        tweet_ids = [result.group(4) for result in results]
+
+        if tweet_ids:
+            confirm = await SimpleConfirm(message, emoji=INSPECT_EMOJI).prompt(ctx)
+            if confirm:
+                async with ctx.typing():
+                    for tweet_id in tweet_ids:
+                        tweet = await self.get_tweet(ctx, tweet_id)
+                        if tweet:
+                            if "extended_entities" in tweet:
```
Comment (noahkw) [inline cogs/Twitter.py:682]:
```suggestion
                        if tweet and "extended_entities" in tweet:
```

File: cogs/Twitter.py [hunk 19]
```diff
@@ -0,0 +1,693 @@
...
+                    channel_list.append(server_settings[0].default_channel)
+                else:
+                    channel_list.extend(servers_dict[server_id])
+            else:
+                channel_list.extend(servers_dict[server_id])
+
+        return channel_list
+
+    async def create_post(self, tweet, is_stream=True):
+        KR_UTC_OFFSET = timedelta(hours=9)
```
Comment (noahkw) [inline cogs/Twitter.py:605]: Put this in line 50

File: cogs/Twitter.py [hunk 20]
```diff
@@ -0,0 +1,693 @@
...
+        except (peony.exceptions.StatusNotFound, peony.exceptions.HTTPNotFound):
+            await ctx.send("Tweet not found")
+            return None
+        return tweet
+
+    async def get_servers(self, session, server_list, tweet_txt):
+        new_server_list = []
+        for server in server_list:
+            server_filters = await db.get_twitter_filters(session, guild_id=server)
+            server_settings = await db.get_twitter_settings(session, server)
```
Comment (noahkw) [inline cogs/Twitter.py:566]: Doing queries in a for loop is a bad idea. Should pass the list of servers to get_twitter_filters/get_twitter_settings.

File: cogs/Twitter.py [hunk 21]
```diff
@@ -0,0 +1,693 @@
...
+            include_entities=True,
+            count=10,
+            tweet_mode="extended",
+        )
+
+        for tweet in timeline:
+            await self.manage_twt(tweet, guild_id=ctx.guild.id)
+
+    async def manage_twt(self, tweet, guild_id=None):
+        async with self.bot.Session() as session:
```
Comment (noahkw) [inline cogs/Twitter.py:515]: Pass a session instead of creating one here

File: cogs/Twitter.py [hunk 22]
```diff
@@ -0,0 +1,693 @@
...
+            screen_name=account,
+            include_rts=False,
+            exclude_replies=True,
+            include_entities=True,
+            count=10,
+            tweet_mode="extended",
+        )
+
+        for tweet in timeline:
+            await self.manage_twt(tweet, guild_id=ctx.guild.id)
```
Comment (noahkw) [inline cogs/Twitter.py:512]: lol

**Review by noahkw — CHANGES_REQUESTED** [review]
Thanks for the PR <3

File: db.py (hunk 2 above)
Comment (chaseemory) [reply db.py:102]: it do

**Review by noahkw — APPROVED** [review]
looks good to me

**Pull request closed without merging**